File: javascript/packages/fury/lib/internalSerializer/datetime.ts
/*
 * Copyright 2023 The Fury Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import  {Fury} from "../type";
import { InternalSerializerType, RefFlags } from "../type";

const epoch =  new Date('1970/01/01 00:00').getTime();

export const timestampSerializer = (fury: Fury) => {
    const { binaryView, binaryWriter} = fury;
    const { writeInt8, writeInt64, writeInt16 } = binaryWriter;
    const { readInt64} = binaryView;
    return {
        read: () => {
            return new Date(Number(readInt64()));
        },
        write: (v: Date) => {
            writeInt8(RefFlags.NotNullValueFlag);
            writeInt16(InternalSerializerType.TIMESTAMP);
            writeInt64(BigInt(v.getTime()));
        },
        config: () => {
            return {
                reserve: 11
            }
        }
    }
}

export const dateSerializer = (fury: Fury) => {
    const { binaryView, binaryWriter} = fury;
    const { writeInt8, writeInt32, writeInt16 } = binaryWriter;
    const { readInt32} = binaryView;
    return {
        read: () => {
            const day = readInt32();
            return new Date(epoch + (day * (24*60*60) * 1000));
        },
        write: (v: Date) => {
            const diff = v.getTime() - epoch;
            const day = Math.floor(diff / 1000 / (24*60*60))
            writeInt8(RefFlags.NotNullValueFlag);
            writeInt16(InternalSerializerType.DATE);
            writeInt32(day);
        },
        config: () => {
            return {
                reserve: 7,
            }
        }
    }
}

File: rust/fory-core/src/serializer/datetime.rs
// Licensed to the Apache Software Foundation (ASF) under one
// or more contributor license agreements.  See the NOTICE file
// distributed with this work for additional information
// regarding copyright ownership.  The ASF licenses this file
// to you under the Apache License, Version 2.0 (the
// "License"); you may not use this file except in compliance
// with the License.  You may obtain a copy of the License at
//
//   http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing,
// software distributed under the License is distributed on an
// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
// KIND, either express or implied.  See the License for the
// specific language governing permissions and limitations
// under the License.

use crate::error::Error;
use crate::fory::Fory;
use crate::resolver::context::ReadContext;
use crate::resolver::context::WriteContext;
use crate::serializer::Serializer;
use crate::types::{FieldType, ForyGeneralList};
use crate::util::EPOCH;
use anyhow::anyhow;
use chrono::{DateTime, Days, NaiveDate, NaiveDateTime};
use std::mem;

impl Serializer for NaiveDateTime {
    fn read(context: &mut ReadContext) -> Result<Self, Error> {
        let timestamp = context.reader.u64();
        DateTime::from_timestamp_millis(timestamp as i64)
            .map(|dt| dt.naive_utc())
            .ok_or(Error::from(anyhow!(
                "Date out of range, timestamp:{timestamp}"
            )))
    }

    fn write(&self, context: &mut WriteContext) {
        context.writer.u64(self.and_utc().timestamp_millis() as u64);
    }

    fn reserved_space() -> usize {
        mem::size_of::<u64>()
    }

    fn get_type_id(_fory: &Fory) -> i16 {
        FieldType::TIMESTAMP.into()
    }
}

impl ForyGeneralList for NaiveDateTime {}

impl Serializer for NaiveDate {
    fn write(&self, context: &mut WriteContext) {
        let days_since_epoch = self.signed_duration_since(EPOCH).num_days();
        context.writer.u64(days_since_epoch as u64);
    }

    fn reserved_space() -> usize {
        mem::size_of::<u64>()
    }

    fn read(context: &mut ReadContext) -> Result<Self, Error> {
        let days = context.reader.u64();
        EPOCH
            .checked_add_days(Days::new(days))
            .ok_or(Error::from(anyhow!(
                "Date out of range, {days} days since epoch"
            )))
    }

    fn get_type_id(_fory: &Fory) -> i16 {
        FieldType::DATE.into()
    }
}

impl ForyGeneralList for NaiveDate {}
